User specific iNat Import API Key

JoeCohen opened the pull request:
Employ a special-purpose, personal MO APIKey for iNat Imports.
- Creates the key if it doesn't already exist.
- Uses the key for all API2 calls during the import.
  - Imports Image, adding it to Observation in a single step, thereby avoiding 2 writes to the db and some emails.
- All objects created during an iNat import are now owned by an MO User other than `webmaster`.
- Removes last vestiges of  `inat_manager`.

### Suggested Manual Test
- Delete any of your [MO API keys](http://localhost:3000/account/api_keys) whose Notes == `inat import`
- Create an iNat observation which has a “Provisional Species Name” Observation Field, with a simple provisional name -- _genus “sp”_  (including the quotes) which does not exist in MO.  Ex: `Mycena "skamaniac"`
See https://www.inaturalist.org/observations/253297232
- Start Solid Queue is running: `bundle exec rake solid_queue:start`
- Import the iNat observation to MO. 
- When importing is Done, click `Results`. (It should display the new Observation with a new provisional Name.)
- Go the Name page (click the Name), scroll to the bottom
Expected Result: The Name should have been Created By you.
(In main, it would be Created and Updated by webmaster.) 
main: <img width="568" alt="Screenshot 2025-04-05 at 2 17 28 PM" src="https://github.com/user-attachments/assets/6950a465-5bb1-4543-bb3e-e46f4bac4215" />

- Go to your [MO API keys](http://localhost:3000/account/api_keys). http://localhost:3000/account/api_keys
Expected Result: You should have a key, created today, with Notes `inat import`.

JoeCohen committed:
Create user inat_import APIKey if user has none

JoeCohen committed:
change method name

JoeCohen committed:
Remove unneeded use of User.current

JoeCohen committed:
Revise test to use different user

JoeCohen committed:
Again change user in test

- It's convenient to use mary for test_create_strip_inat_username because she lacks APIKey for inat imports
  - cannot use rolf because he has a key because he's the default user for InatImportJobTest
  - cannot use dick because he has a key because he's a superuser in InatImportJobTest

JoeCohen committed:
Use user's personal APIKey instead of inat_manager's

JoeCohen committed:
Merge main' into jdc-2849-inat-individual-api-key, resolving conflicts

JoeCohen committed:
Fix assertion msgs

JoeCohen committed:
Remove inat_manager

This role is obsolete;
all objects created during import have a different user.

JoeCohen committed:
add comment

JoeCohen committed:
Extract method in order to make RuboCop happy

JoeCohen committed:
Delete empty line befor rescue.

It offends RuboCop

JoeCohen committed:
Delete obsolete inat_manager_key method

JoeCohen committed:
fix alignmentoffense

JoeCohen committed:
Fix api_keys fixtures

- Add defaults, making keys unique

JoeCohen committed:
Remove updating of added sequences

No longer needed by sequence is created with user = @user, rather than inat_manager

JoeCohen committed:
Delete one more stray inat_manager

JoeCohen committed:
Verify user's iNat import APIKey

Fixes bug in this branch where I created a key, but didn't verify it.
That makes API return errors.

JoeCohen committed:
Test ownership of imported images

JoeCohen committed:
Use api to add photo in a single step

- Revises params used by the api
- Deletes now-obsolete code.
By using the importomg user's api key and the correct params above,
the api imports the image with the correct attributes and adds it to the Observation.
So there's no need to update the image.

JoeCohen committed:
Change local ariable name or clarity

JoeCohen committed:
Remove lingering vestiges of inat_import_manager

JoeCohen committed:
Fix test  assertion & message

How did this ever work?

JoeCohen committed:
Add stubbish hack to job to fake side-effect of api call

JoeCohen commented on app/jobs/inat_import_job.rb:
```diff
@@ -255,42 +260,35 @@ def add_inat_images(inat_obs_photos)
-        user_id: @user.id,
-        when: @observation.when,
-        original_name: photo.original_name
+      next unless Rails.env.test?
+
+      # NOTE: 2025-04-01 jdc
+      # A hack faking the side effect of the above API call
+      # The API adds the imported image to the MO observation
+      image = api.results.first
+      ObservationImage.create!(
+        observation: @observation,
+        image: image
```
I can't figure out how to stub, mock, or otherwise fake this side effect in the test (`test/jobs/inat_import_job_test.rb`).

`Inat::PhotoImporter` wraps an API call.
The API imports an image and **adds it to an MO Observation**.
The tests do mock `Inat::PhotoImporter` and return a value that's sufficient for testing.
When the mock is created the Observation doesn't exist, so I don't know how tell the mock which Observation needs to have the image added.

The hacky code in InatImportJob was the best I could do.
When that code is run both the Observation and the (mock) Image are knowm.

mo-nathan replied:
Start looking at this, but need some more time.

mo-nathan replied:
Ok, I found the issue.  The primary issue was that you weren't actually providing the system with an image to connect to the observation.  I have version of the code on njw-inat-job-stub-fix that removes the dependency on Rails.env.test? and all the tests pass.  This required removing some of the `Inat::PhotoImporter.stub`s.  I actually wonder if all of those should get removed, but I didn't go that far.  The key new function is `image_for_stubs` which provides an actual image in the stub response to the iNat get calls.  I spread this around liberally.  What I didn't do was refactor the new stub calls to keep them DRY.  Finally, I ran into an issue where `imported_img.original_name` was returning `nil` rather than the expected value.  This turns out to be due to the user `rolf` having the keep_filenames preference set to `toss`.  I don't know how important that bit of the test is to you, so I just covered ith with a simple guard condition.  So in summary, before merging that PR into this PR, I recommend:
1. Reviewin all the stubbings of `Inat::PhotoImporter.stub` and see if you can replace them with lower level stubs to the iNat get calls.  I think this will conceptually at least simplify things and it is a better place to do the stubbing.
2. Look at refactoring the iNat get calls into one or two helpful functions.  You have started down this road with `stub_inat_photo_requests`, but there seem to be two different places that the stored iNat responses expect their images to come from (`https://inaturalist-open-data.s3.amazonaws.com/photos` and `https://static.inaturalist.org/photos/`).  Not sure why, but it's what I observed while working on the tests.
3. Review the conditional I put on `imported_img.original_name` and figure out what you really want there.  Maybe you want to use an other users, maybe you want to not have that assert, maybe something else.

JoeCohen committed:
Test added names

- Proves
  - MO added:
    - a species Name corresponding to the iNat name
    - a genus corresponding to the above
  - The author of those names is the importing user (rather than webmaster).

JoeCohen committed:
Fix typo in comment

JoeCohen committed:
Add some comments

JoeCohen committed:
Remove stray breakpoint

mo-nathan commented:
Was able to confirm that it works from local.

mo-nathan committed:
Merge branch 'main' into jdc-2849-inat-individual-api-key

mo-nathan committed:
Merge branch 'main' into jdc-2849-inat-individual-api-key

JoeCohen committed:
Re-create original PR2889

Re-creates #2889 without my messed up commits and pushes

JoeCohen committed:
Fix cannot create regular file Errors

See https://github.com/MushroomObserver/mushroom-observer/actions/runs/14364409885/job/40273711486?pr=2890

JoeCohen committed:
Delete stray breakpoint

JoeCohen committed:
Rubocop

JoeCohen committed:
Restore assertion of original_name of imported photo

- Responds to @mo-nathan's point 3 in https://github.com/MushroomObserver/mushroom-observer/pull/2851#discussion_r2033906999
- Removes conditional which skipped the assertion
- Add a new user fixture to serve as the default user in tests
- adjust tests

JoeCohen committed:
Rubocop useless assignment to variable

JoeCohen committed:
Merge pull request #2890 from MushroomObserver/new-njw-remove-rails-env-test-dependency

Remove env.test? dependency

JoeCohen committed:
Shorten instance variable name

JoeCohen committed:
Add comment about non-CC license iNat photos

JoeCohen committed:
Revise stub_inat_photo_requests to work with unlicensed photos

JoeCohen committed:
Clarify intent of test method

- Add comment explaining method raison d'etre
- Changes method name
- Corrects comment

JoeCohen committed:
Reorganize file to move stubs to another file

- Gathers stubs code
- Deletes unused method

JoeCohen committed:
Remove more unused methods

JoeCohen committed:
Revise stub constants

- Adds clarifying comments
- Add constant for unlicensed photo source

JoeCohen committed:
Promote import_job_obs_with_one_photo to a test

It had been called from from another test (test_import_job_obs_with_sequence_and_multiple_ids)
because it failed on its own.
Nathan's various `stub_request`'s cured that.

JoeCohen committed:
Remove calls to     Inat::PhotoImporter.stub

JoeCohen committed:
Fix bug in stub_inat_photo_requests

- Changes how photo url is derived, thereby allowing different photo file extensions
- DRYs unneeded different methods for CC-licensed vs unlicensed photos

JoeCohen committed:
Same logic in tests and app for iNat original image url

JoeCohen committed:
Remove remaining individual GET photo stubs

JoeCohen committed:
Remove now-unneeded image_stubs method

JoeCohen committed:
Remove stray comment

JoeCohen committed:
Remove now-unneeded stubs of Inat::PhotoImporter

JoeCohen committed:
Toss funky now-unneeded test mocking of Inat::PhotoImporter

Testing importing photos is now stubbed at a lower level.

JoeCohen committed:
Remove unneeded InatPhotoImporter

It's no longer needed.
It existed only because I couldn't stub the API's get requests.
That's now accomplished (thanks to Nathan).

JoeCohen committed:
Improve comment

JoeCohen committed:
Delete confusing comment

JoeCohen committed:
Improve method name

mo-nathan reviewed (approved):
Looks good!
